app/src/main/java/fr/free/nrw/commons/upload/FailedUploadsFragment.kt

```diff
@@ -0,0 +1,199 @@
+        pendingUploadsPresenter!!.getFailedContributions(userName)
+        pendingUploadsPresenter!!.failedContributionList.observe(
+            viewLifecycleOwner
+        ) { list: PagedList<Contribution?> ->
+            contributionsList = ArrayList()
+            list.forEach {
+                if (it != null) {
+                    contributionsList.add(it)
+                }
+            }
+            if (list.size == 0) {
+                uploadProgressActivity.setErrorIconsVisibility(false)
+                binding.nofailedTextView.visibility = View.VISIBLE
+                binding.failedUplaodsLl.visibility = View.GONE
+            } else {
+                uploadProgressActivity.setErrorIconsVisibility(true)
+                binding.nofailedTextView.visibility = View.GONE
+                binding.failedUplaodsLl.visibility = View.VISIBLE
+                val adapter = FailedUploadsAdapter(contributionsList, this)
+                binding.failedUploadsRecyclerView.setAdapter(adapter)
+            }
+        }
+    }
+
+    companion object {
+        @JvmStatic
+        fun newInstance(param1: String, param2: String) =
+            FailedUploadsFragment().apply {
+                arguments = Bundle().apply {
+                    putString(ARG_PARAM1, param1)
+                    putString(ARG_PARAM2, param2)
+                }
+            }
+    }
+
+    fun restartUploads() {
+        if (contributionsList != null) {
+            uploadProgressActivity.resetProgressBar()
+            pendingUploadsPresenter.restartUploads(
+                contributionsList,
+                0,
+                this.requireContext().applicationContext
+            )
+        }
+    }
+
+    override fun restartUpload(index : Int) {
+        if (contributionsList != null) {
+            uploadProgressActivity.resetProgressBar()
+            pendingUploadsPresenter.restartUpload(
+                contributionsList,
+                index,
+                this.requireContext().applicationContext
+            )
+        }
+    }
+
+    override fun deleteUpload(contribution: Contribution?) {
+        DialogUtil.showAlertDialog(
+            requireActivity(),
+            String.format(
+                Locale.getDefault(),
+                getString(R.string.cancelling_upload)
+            ),
+            String.format(
+                Locale.getDefault(),
+                getString(R.string.cancel_upload_dialog)
+            ),
+            String.format(Locale.getDefault(), getString(R.string.yes)),
+            String.format(Locale.getDefault(), getString(R.string.no)),
+            {
+                ViewUtil.showShortToast(context, R.string.cancelling_upload)
+                pendingUploadsPresenter.deleteUpload(
+                    contribution,
+                    this.requireContext().applicationContext
+                )
+                CommonsApplication.cancelledUploads.add(contribution!!.pageId)
```

Comment:
Does our new queue get updated in real time?

Reply:
It do gets updated in the real time

Reply:
Do we really require this hashset then? It was introduced when we where using a one-shot query. If the real queue handles this case, we could remove it.